lib/wasi/src/bin_factory/binary_package.rs

```diff
@@ -0,0 +1,135 @@
+use std::{
+    any::Any,
+    borrow::Cow,
+    collections::HashMap,
+    sync::{Arc, Mutex, RwLock},
+};
+
+use derivative::*;
+use wasmer_vfs::{FileSystem, TmpFileSystem};
+use wasmer_wasi_types::wasi::Snapshot0Clockid;
+
+use super::hash_of_binary;
+use crate::syscalls::platform_clock_time_get;
+
+#[derive(Derivative, Clone)]
+#[derivative(Debug)]
+pub struct BinaryPackageCommand {
+    pub name: String,
+    #[derivative(Debug = "ignore")]
+    pub atom: Cow<'static, [u8]>,
+    hash: Option<String>,
+    pub ownership: Option<Arc<dyn Any + Send + Sync + 'static>>,
+}
+
+impl BinaryPackageCommand {
+    pub fn new(name: String, atom: Cow<'static, [u8]>) -> Self {
+        Self {
+            name,
+            ownership: None,
+            hash: None,
+            atom,
+        }
+    }
+
+    pub unsafe fn new_with_ownership<'a, T>(
+        name: String,
+        atom: Cow<'a, [u8]>,
+        ownership: Arc<T>,
+    ) -> Self
+    where
+        T: 'static,
+    {
+        let ownership: Arc<dyn Any> = ownership;
+        let mut ret = Self::new(name, std::mem::transmute(atom));
+        ret.ownership = Some(std::mem::transmute(ownership));
+        ret
+    }
+
+    pub fn hash(&mut self) -> &str {
+        if self.hash.is_none() {
+            self.hash = Some(hash_of_binary(self.atom.as_ref()));
+        }
+        let hash = self.hash.as_ref().unwrap();
+        hash.as_str()
+    }
+}
+
+#[derive(Derivative, Clone)]
+#[derivative(Debug)]
+pub struct BinaryPackage {
+    pub package_name: Cow<'static, str>,
+    pub when_cached: Option<u128>,
+    pub ownership: Option<Arc<dyn Any + Send + Sync + 'static>>,
+    #[derivative(Debug = "ignore")]
+    pub entry: Option<Cow<'static, [u8]>>,
+    pub hash: Arc<Mutex<Option<String>>>,
+    pub wapm: Option<String>,
```

Comment:
What is this and how it differs from `package_name`?

Reply:
I'm working on more or less completely rewriting this anyway.

Reply:
Part of #3605